File: lib/board.rb
require_relative "square"

class Board

  def initialize
    make_squares
  end

  def all
    unless @all
      @all = @squares.values.flatten.map { |hsh| hsh.values[0] }
    end
    @all
  end

  def find hsh
    if @squares[hsh[:x]] && @squares[hsh[:x]][hsh[:y]]
      @squares[hsh[:x]][hsh[:y]]
    end
  end

private
  
  def make_squares
    @squares = {}
    (0..8).each do |x|
      @squares[x] = {}
      (0..9).each do |y|
        square = Square.new x, y
        @squares[x][y] = square
        if @squares[x - 1]
          left_square = @squares[x - 1][y]
          square.point left: left_square
          left_square.point right: square
        end
        if @squares[x][y - 1]
          down_square = @squares[x][y - 1]
          square.point down: down_square
          down_square.point up: square
        end
      end
    end
  end
end

File: src/board.js
var Square = require('./square.js');
module.exports = Board;

function Board() {

  var continents = { red: {}, black: {} };
  var palaces = { red: {}, black: {} };

  for (var x = 0; x < 9; x += 1) {
    for (var y = 0; y < 10; y += 1) {
      var coordinates = x + ',' + y;
      var square = new Square(x, y);
      this[coordinates] = square;
      var leftSquare = this[(x - 1) + ',' + y];
      var downSquare = this[x + ',' + (y - 1)];
      if (leftSquare) { square.left = leftSquare; leftSquare.right = square; }
      if (downSquare) { square.down = downSquare; downSquare.up = square }
      if (y < 5) {
        continents.red[coordinates] = square;
      } else {
        continents.black[coordinates] = square;
      }
      if (y < 3 && x < 6 && x > 2) {
        palaces.red[coordinates] = square;
      } else if (y > 6 && x < 6 && x > 2) {
        palaces.black[coordinates] = square;
      }
    }
  }

  Object.defineProperties(this, {
    continents: { value: continents, enumerable: false },
    palaces: { value: palaces, enumerable: false }
  });
}
